File: sdk/swift/Sources/OpenEye/OpenEyeClient.swift
import Foundation

/// HTTP client for the OpenEye sidecar. Mirrors the TypeScript `SidecarClient`
/// API surface for iOS / visionOS apps.
///
/// The Python sidecar is expected to be reachable over the network — typically
/// running on a workstation or a service in the same WiFi/LAN. iOS apps cannot
/// spawn the Python process themselves; you run the sidecar separately and
/// point this client at it.
///
/// Quick start:
///
/// ```swift
/// let client = OpenEyeClient(baseURL: URL(string: "http://192.168.1.50:7770")!)
/// let session = try await client.createSession(source: "ios-app")
/// let visualSession = try await client.createVisualSession(deviceType: "visionos")
/// try await client.logFrame(
///     visualSessionId: visualSession,
///     sequenceNum: 1,
///     sceneDescription: "operator placing bolt"
/// )
/// try await client.logStepVerification(
///     visualSessionId: visualSession,
///     stepId: "s1",
///     result: .pass
/// )
/// ```
public final class OpenEyeClient: @unchecked Sendable {

    public enum OpenEyeError: Error {
        case httpError(status: Int, body: String)
        case decodingError(underlying: Error)
        case missingResponse
    }

    public enum VerifyResult: String, Codable {
        case pass, fail, uncertain
    }

    private let baseURL: URL
    private let session: URLSession
    private let token: String?
    private let requestTimeout: TimeInterval

    public init(
        baseURL: URL,
        sidecarToken: String? = nil,
        requestTimeout: TimeInterval = 15.0,
        urlSession: URLSession = .shared
    ) {
        self.baseURL = baseURL
        self.token = sidecarToken
        self.requestTimeout = requestTimeout
        self.session = urlSession
    }

    // MARK: - Internal helpers

    func request(_ method: String, _ path: String, body: [String: Any]? = nil) async throws -> Data {
        var req = URLRequest(url: baseURL.appendingPathComponent(path))
        req.httpMethod = method
        req.timeoutInterval = requestTimeout
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = token {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body = body {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw OpenEyeError.missingResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw OpenEyeError.httpError(status: http.statusCode, body: body)
        }
        return data
    }

    func decode<T: Decodable>(_ data: Data, as _: T.Type) throws -> T {
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw OpenEyeError.decodingError(underlying: error)
        }
    }

    // MARK: - Health

    public func health() async throws -> Bool {
        struct Health: Decodable { let ok: Bool }
        let data = try await request("GET", "health")
        return try decode(data, as: Health.self).ok
    }

    // MARK: - Sessions

    public func createSession(
        source: String = "ios",
        userId: String? = nil,
        tenantId: String? = nil,
        model: String? = nil,
        systemPrompt: String? = nil
    ) async throws -> String {
        struct Response: Decodable { let session_id: String }
        var body: [String: Any] = ["source": source]
        body["user_id"] = userId
        body["tenant_id"] = tenantId
        body["model"] = model
        body["system_prompt"] = systemPrompt
        let data = try await request("POST", "sessions/create", body: body.compactMapValues { $0 })
        return try decode(data, as: Response.self).session_id
    }

    public func endSession(_ sessionId: String, reason: String = "completed") async throws {
        _ = try await request("POST", "sessions/\(sessionId)/end", body: ["reason": reason])
    }

    public func appendMessage(
        sessionId: String,
        role: String,
        content: String? = nil
    ) async throws -> Int {
        struct Response: Decodable { let message_id: Int }
        var body: [String: Any] = ["role": role]
        body["content"] = content
        let data = try await request("POST", "sessions/\(sessionId)/messages",
                                     body: body.compactMapValues { $0 })
        return try decode(data, as: Response.self).message_id
    }

    // MARK: - Visual sessions

    public func createVisualSession(
        deviceType: String,
        deviceId: String? = nil,
        procedureId: String? = nil,
        procedureName: String? = nil,
        userId: String? = nil,
        tenantId: String? = nil,
        sessionId: String? = nil
    ) async throws -> String {
        struct Response: Decodable { let visual_session_id: String }
        var body: [String: Any] = ["device_type": deviceType]
        body["device_id"] = deviceId
        body["procedure_id"] = procedureId
        body["procedure_name"] = procedureName
        body["user_id"] = userId
        body["tenant_id"] = tenantId
        body["session_id"] = sessionId
        let data = try await request("POST", "visual-sessions/create",
                                     body: body.compactMapValues { $0 })
        return try decode(data, as: Response.self).visual_session_id
    }

    public func endVisualSession(_ vsid: String, outcome: String = "completed") async throws {
        _ = try await request("POST", "visual-sessions/\(vsid)/end",
                              body: ["outcome": outcome])
    }

    // MARK: - Frames

    public func logFrame(
        visualSessionId: String,
        sequenceNum: Int,
        sceneDescription: String,
        objectsDetected: [String]? = nil,
        stepContext: String? = nil,
        confidence: Double? = nil
    ) async throws -> Int {
        struct Response: Decodable { let frame_id: Int }
        var body: [String: Any] = [
            "visual_session_id": visualSessionId,
            "sequence_num": sequenceNum,
            "scene_description": sceneDescription,
        ]
        body["objects_detected"] = objectsDetected
        body["step_context"] = stepContext
        body["confidence"] = confidence
        let data = try await request("POST", "frames/log",
                                     body: body.compactMapValues { $0 })
        return try decode(data, as: Response.self).frame_id
    }

    // MARK: - Step verification

    public func logStepVerification(
        visualSessionId: String,
        stepId: String,
        result: VerifyResult,
        stepName: String? = nil,
        confidence: Double? = nil,
        reasoning: String? = nil,
        frameId: Int? = nil
    ) async throws -> Int {
        struct Response: Decodable { let verification_id: Int }
        var body: [String: Any] = [
            "visual_session_id": visualSessionId,
            "step_id": stepId,
            "result": result.rawValue,
        ]
        body["step_name"] = stepName
        body["confidence"] = confidence
        body["reasoning"] = reasoning
        body["frame_id"] = frameId
        let data = try await request("POST", "steps/log",
                                     body: body.compactMapValues { $0 })
        return try decode(data, as: Response.self).verification_id
    }
}

File: sdk/kotlin/src/main/kotlin/info/getcontext/openeye/OpenEyeClient.kt
package info.getcontext.openeye

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.sse.EventSource
import okhttp3.sse.EventSourceListener
import okhttp3.sse.EventSources
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * HTTP client for the OpenEye sidecar — Kotlin/Android edition.
 *
 * Mirrors the TypeScript SidecarClient and Swift OpenEyeClient API.
 *
 * IMPORTANT: Android apps cannot spawn the Python sidecar. Run the sidecar
 * on a server or workstation, then point this client at it over HTTPS in
 * production or HTTP on a LAN for development.
 *
 * Quick start:
 * ```kotlin
 * val client = OpenEyeClient("http://192.168.1.50:7770")
 * val sessionId = client.createSession(source = "android-app")
 * val vsid = client.createVisualSession(deviceType = "android")
 * client.logFrame(vsid, sequenceNum = 1, sceneDescription = "operator placing bolt")
 * client.logStepVerification(vsid, stepId = "s1", result = VerifyResult.PASS)
 * ```
 */
class OpenEyeClient(
    private val baseUrl: String,
    private val sidecarToken: String? = null,
    private val requestTimeoutSeconds: Long = 15,
    private val httpClient: OkHttpClient = defaultClient(requestTimeoutSeconds),
) {
    enum class VerifyResult(val value: String) {
        PASS("pass"),
        FAIL("fail"),
        UNCERTAIN("uncertain"),
    }

    class OpenEyeHttpException(val status: Int, val body: String) :
        RuntimeException("OpenEye HTTP $status: $body")

    companion object {
        private val JSON_MEDIA = "application/json".toMediaType()
        private val json = Json { ignoreUnknownKeys = true }

        private fun defaultClient(timeoutSeconds: Long) = OkHttpClient.Builder()
            .connectTimeout(timeoutSeconds, TimeUnit.SECONDS)
            .readTimeout(timeoutSeconds, TimeUnit.SECONDS)
            .writeTimeout(timeoutSeconds, TimeUnit.SECONDS)
            // SSE streams need no read timeout
            .build()
    }

    private fun urlFor(path: String): String =
        "${baseUrl.trimEnd('/')}/${path.trimStart('/')}"

    private fun buildRequest(method: String, path: String, body: Any? = null): Request {
        val builder = Request.Builder().url(urlFor(path))
        sidecarToken?.let { builder.header("Authorization", "Bearer $it") }
        when (method.uppercase()) {
            "GET" -> builder.get()
            "POST" -> {
                val payload = if (body == null) {
                    "".toRequestBody(JSON_MEDIA)
                } else {
                    json.encodeToString(JsonObject.serializer(),
                        body as JsonObject).toRequestBody(JSON_MEDIA)
                }
                builder.post(payload)
            }
            else -> throw IllegalArgumentException("Unsupported method: $method")
        }
        return builder.build()
    }

    private suspend fun executeForJson(request: Request): JsonObject =
        suspendCancellableCoroutine { cont ->
            val call = httpClient.newCall(request)
            cont.invokeOnCancellation { call.cancel() }
            call.enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    cont.resumeWithException(e)
                }
                override fun onResponse(call: Call, response: Response) {
                    response.use {
                        val bodyText = it.body?.string() ?: ""
                        if (!it.isSuccessful) {
                            cont.resumeWithException(
                                OpenEyeHttpException(it.code, bodyText))
                            return
                        }
                        val parsed = if (bodyText.isBlank()) JsonObject(emptyMap())
                            else json.parseToJsonElement(bodyText).jsonObject
                        cont.resume(parsed)
                    }
                }
            })
        }

    private fun jsonObjectOf(vararg pairs: Pair<String, JsonElement?>): JsonObject =
        JsonObject(pairs.mapNotNull { (k, v) -> if (v == null) null else k to v }.toMap())

    // MARK: - Health

    suspend fun health(): Boolean {
        val resp = executeForJson(buildRequest("GET", "health"))
        return (resp["ok"] as? kotlinx.serialization.json.JsonPrimitive)?.content?.toBooleanStrictOrNull() == true
    }

    // MARK: - Sessions

    suspend fun createSession(
        source: String = "android",
        userId: String? = null,
        tenantId: String? = null,
        model: String? = null,
    ): String {
        val body = jsonObjectOf(
            "source" to kotlinx.serialization.json.JsonPrimitive(source),
            "user_id" to userId?.let { kotlinx.serialization.json.JsonPrimitive(it) },
            "tenant_id" to tenantId?.let { kotlinx.serialization.json.JsonPrimitive(it) },
            "model" to model?.let { kotlinx.serialization.json.JsonPrimitive(it) },
        )
        val resp = executeForJson(buildRequest("POST", "sessions/create", body))
        return resp["session_id"]!!.jsonPrimitive.content
    }

    suspend fun endSession(sessionId: String, reason: String = "completed") {
        val body = jsonObjectOf("reason" to kotlinx.serialization.json.JsonPrimitive(reason))
        executeForJson(buildRequest("POST", "sessions/$sessionId/end", body))
    }

    // MARK: - Visual sessions

    suspend fun createVisualSession(
        deviceType: String,
        deviceId: String? = null,
        procedureId: String? = null,
        procedureName: String? = null,
        tenantId: String? = null,
        sessionId: String? = null,
    ): String {
        val body = jsonObjectOf(
            "device_type" to kotlinx.serialization.json.JsonPrimitive(deviceType),
            "device_id" to deviceId?.let { kotlinx.serialization.json.JsonPrimitive(it) },
            "procedure_id" to procedureId?.let { kotlinx.serialization.json.JsonPrimitive(it) },
            "procedure_name" to procedureName?.let { kotlinx.serialization.json.JsonPrimitive(it) },
            "tenant_id" to tenantId?.let { kotlinx.serialization.json.JsonPrimitive(it) },
            "session_id" to sessionId?.let { kotlinx.serialization.json.JsonPrimitive(it) },
        )
        val resp = executeForJson(buildRequest("POST", "visual-sessions/create", body))
        return resp["visual_session_id"]!!.jsonPrimitive.content
    }

    suspend fun endVisualSession(vsid: String, outcome: String = "completed") {
        val body = jsonObjectOf("outcome" to kotlinx.serialization.json.JsonPrimitive(outcome))
        executeForJson(buildRequest("POST", "visual-sessions/$vsid/end", body))
    }

    // MARK: - Frames

    suspend fun logFrame(
        visualSessionId: String,
        sequenceNum: Int,
        sceneDescription: String,
        stepContext: String? = null,
        confidence: Double? = null,
    ): Int {
        val body = jsonObjectOf(
            "visual_session_id" to kotlinx.serialization.json.JsonPrimitive(visualSessionId),
            "sequence_num" to kotlinx.serialization.json.JsonPrimitive(sequenceNum),
            "scene_description" to kotlinx.serialization.json.JsonPrimitive(sceneDescription),
            "step_context" to stepContext?.let { kotlinx.serialization.json.JsonPrimitive(it) },
            "confidence" to confidence?.let { kotlinx.serialization.json.JsonPrimitive(it) },
        )
        val resp = executeForJson(buildRequest("POST", "frames/log", body))
        return resp["frame_id"]!!.jsonPrimitive.content.toInt()
    }

    // MARK: - Step verification

    suspend fun logStepVerification(
        visualSessionId: String,
        stepId: String,
        result: VerifyResult,
        stepName: String? = null,
        confidence: Double? = null,
        reasoning: String? = null,
    ): Int {
        val body = jsonObjectOf(
            "visual_session_id" to kotlinx.serialization.json.JsonPrimitive(visualSessionId),
            "step_id" to kotlinx.serialization.json.JsonPrimitive(stepId),
            "result" to kotlinx.serialization.json.JsonPrimitive(result.value),
            "step_name" to stepName?.let { kotlinx.serialization.json.JsonPrimitive(it) },
            "confidence" to confidence?.let { kotlinx.serialization.json.JsonPrimitive(it) },
            "reasoning" to reasoning?.let { kotlinx.serialization.json.JsonPrimitive(it) },
        )
        val resp = executeForJson(buildRequest("POST", "steps/log", body))
        return resp["verification_id"]!!.jsonPrimitive.content.toInt()
    }

    // MARK: - SSE stream

    data class OpenEyeEvent(
        val type: String,
        val sessionId: String?,
        val timestamp: Double,
        val data: JsonObject,
    )

    /** Flow of events from /sessions/{sessionId}/events. Use "*" for all sessions. */
    fun events(sessionId: String): Flow<OpenEyeEvent> = callbackFlow {
        val request = buildRequest("GET", "sessions/$sessionId/events")
        val listener = object : EventSourceListener() {
            override fun onEvent(eventSource: EventSource, id: String?, type: String?, data: String) {
                try {
                    val obj = json.parseToJsonElement(data).jsonObject
                    val evt = OpenEyeEvent(
                        type = obj["type"]?.jsonPrimitive?.contentOrNull ?: type ?: "unknown",
                        sessionId = obj["session_id"]?.jsonPrimitive?.contentOrNull,
                        timestamp = obj["ts"]?.jsonPrimitive?.contentOrNull?.toDoubleOrNull() ?: 0.0,
                        data = obj["data"]?.jsonObject ?: JsonObject(emptyMap()),
                    )
                    trySend(evt)
                } catch (_: Exception) {
                    // Drop malformed frame; SSE keeps streaming
                }
            }
            override fun onFailure(eventSource: EventSource, t: Throwable?, response: Response?) {
                close(t ?: RuntimeException("SSE failed: ${response?.code}"))
            }
            override fun onClosed(eventSource: EventSource) { close() }
        }
        val source = EventSources.createFactory(httpClient).newEventSource(request, listener)
        awaitClose { source.cancel() }
    }.flowOn(Dispatchers.IO)
}
